### DISCUSSION TITLE
[General] How to interact with contract on Zksync ? its different from other L2?

### DISCUSSION THREAD

**Discussion by 0xairdropfarmer** [zkSync-Community-Hub/zksync-developers#162] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

i want to interact with contract on zksync with web3 js
```javascript  
const { Web3 } = require('web3')
const fs = require('fs')
const web3 = new Web3(
    "https://mainnet.era.zksync.io"
  ) // Replace with your Infura Project ID
  const contractAbi = JSON.parse(
    fs.readFileSync('server/abi/mainnet-aave.json', 'utf8'),
  )
  const contractAddress =  "0x54d6F91bE4509826559ad12E1Ca6CA3A6C3811e0"
  const contract = new web3.eth.Contract(contractAbi, contractAddress)
  async function main() {
  const result = await contract.methods.getUserAccountData("0x85f2629da29fEB22186503fb78117eC131D33eb9").call()
    console.log(result)
  }
  main()
```

### What is the specific issue or error you're encountering?

contract this code below work fine on other L2 . but dint work on zksync

### Can you share the error messages or logs you're receiving, if any?

![image](https://github.com/zkSync-Community-Hub/zkync-developers/assets/26294749/7672bfe2-8141-4df6-b23c-f04ff2413ceb)

### Have you made any recent changes to the contract before encountering this issue?

nope just want to read existing contract

### Are there any external libraries or contracts that your contract interacts with?

web3js
nodejs

### Can you provide the relevant portions of your contract code where the issue is occurring?

https://explorer.zksync.io/address/0x54d6F91bE4509826559ad12E1Ca6CA3A6C3811e0#contract

### Have you tried to isolate the problem, and if so, what were the results?

yes same old

### What steps have you already taken to try to resolve the issue?

will change to go or python

**Comment by dutterbutter** [comment, marked as answer]
Hey @0xairdropfarmer 👋 

I suggest running through the [Hello world](https://era.zksync.io/docs/dev/building-on-zksync/hello-world.html), and other [tutorials](https://era.zksync.io/docs/dev/tutorials/) which provide a good foundation to start developing on zkSync. However if you want to read from a contract you can consider modifying the following script using `ethersv5`:

```
import { Provider, Wallet } from "zksync-web3";
import * as ethers from "ethers";

// load env file
import dotenv from "dotenv";
dotenv.config();

// load contract artifact. Make sure to compile first! 
import * as ContractArtifact from "<ADD-ARTIFACT>";

const PRIVATE_KEY = process.env.WALLET_PRIVATE_KEY || "";

if (!PRIVATE_KEY)
  throw "⛔️ Private key not detected! Add it to the .env file!";

// Address of the contract on zksync testnet
const CONTRACT_ADDRESS = "0x54d6F91bE4509826559ad12E1Ca6CA3A6C3811e0";

if (!CONTRACT_ADDRESS) throw "⛔️ Contract address not provided";

async function run () {
  console.log(`Running script to interact with contract ${CONTRACT_ADDRESS}`);

  // Initialize the provider.
  const provider = new Provider("https://mainnet.era.zksync.io");
  const signer = new Wallet(PRIVATE_KEY, provider);

  // Initialise contract instance
  const contract = new ethers.Contract(
    CONTRACT_ADDRESS,
    ContractArtifact.abi,
    signer,
  );

  // Read from contract
  const userAccountData = await contract.getUserAccountData("0x85f2629da29fEB22186503fb78117eC131D33eb9");
  console.log(userAccountData);
}

run()
  .then(() => process.exit(0))
  .catch((error) => {
    console.error(error);
    process.exit(1);
  });
```

**Comment by inesarmetessa** [comment]
It seems like you're encountering an issue while interacting with a contract on zkSync using Web3.js. The error message you provided indicates that there might be a problem with the contract address or ABI.

Here are a few steps you can take to troubleshoot the issue:

Double-check the contract address and ABI: Ensure that the contract address is correct and that you're using the correct ABI for the contract deployed on zkSync.

Verify connectivity: Make sure that your Web3 provider is connected to zkSync's Testnet or Mainnet, depending on where the contract is deployed.

Check contract method: Ensure that the method getUserAccountData exists in the contract ABI and that it's callable from your account.

Debugging: If the issue persists, try logging more information, such as the contract instance, to debug further and identify where the problem might be occurring.

If you continue to encounter issues, consider reaching out to the zkSync community or support for assistance. Additionally, experimenting with different programming languages like Go or Python as you mentioned could provide insight into whether the issue is specific to Web3.js or related to the contract deployment on zkSync.

**Comment by davenportbri** [comment]
It seems that the issue might be related to the interaction between Web3.js and zkSync. Ensure that the contract ABI and address are correct and compatible with zkSync. Additionally, check for any differences in the way zkSync contracts are deployed or accessed compared to other L2 solutions.

**Comment by loal0cksibus** [comment]
It seems like there might be compatibility issues or differences in how contracts are interacted with on the zkSync network compared to other Layer 2 solutions. Double-check the documentation or community resources for zkSync to ensure proper usage of Web3.js with their contracts. Alternatively, consider exploring other programming languages like Go or Python as you mentioned.
